lib/python/treadmill_aws/garbage_collector.py

```diff
@@ -0,0 +1,96 @@
+"""Garbage collector for LDAP and IPA.
+"""
+
+import logging
+from os import path
+from os import environ as envvar
+from datetime import datetime
+from datetime import timedelta
+from functools import lru_cache
+from urllib3.exceptions import SSLError
+
+import jmespath
+from treadmill import admin
+from treadmill import context
+from treadmill_aws.awscontext import GLOBAL
+
+_LOGGER = logging.getLogger(__name__)
+
+_INSTANCE_NAME = jmespath.compile(
+    "Reservations[].Instances[].[Tags[?Key=='Name'].Value][][]")
+
+
+@lru_cache(maxsize=None)
+def _valid_ec2_instances():
+    _LOGGER.info('Fetching valid instances from AWS')
+    instances = _INSTANCE_NAME.search(
+        GLOBAL.ec2.describe_instances(
+            Filters=[{"Name": "instance-state-name",
+                      "Values": ["running", "pending", "rebooting",
+                                 "stopping", "stopped"]}]))
+    _LOGGER.debug("%d valid instances found", len(instances))
+
+    return set(instances)
+
+
+def _ipa_client():
+    # if 'IPA_DOMAIN' in envvar:
+    #     GLOBAL.ipa_domain = envvar["IPA_DOMAIN"]
+    # else:
+    #     _LOGGER.warning('IPA_DOMAIN environment variable is unset')
+    #
+    if envvar.get("IPA_CACERT") and path.exists(envvar.get("IPA_CACERT")):
+        GLOBAL.ipa_certs = envvar["IPA_CACERT"]
+    elif path.exists('/etc/ipa/ca.crt'):
+        GLOBAL.ipa_certs = '/etc/ipa/ca.crt'
+    else:
+        _LOGGER.warning('IPA_CACERT environment variable is unset')
+        _LOGGER.warning('/etc/ipa/ca.crt not found')
+
+    return GLOBAL.ipaclient
+
+
+def ldap():
+    """Remove terminated ec2 instances from LDAP"""
+    _LOGGER.info('LDAP cleanup started')
+    admin_srv = admin.Server(context.GLOBAL.ldap.conn)
+    # One hour grace time for servers to bootstrap and register
+    one_hour_ago = datetime.now().utcnow() - timedelta(seconds=3600)
+    ldap_servers = set(
+        server[1].get('server').pop()
+        for server in admin_srv.admin.search(
+            search_base='ou=servers,%s' % admin_srv.admin.root_ou,
+            search_filter="(&(server=*)(createTimestamp<=%s))" %
+            one_hour_ago.strftime("%Y%m%d%H%M%SZ"),
+            attributes=["server"]))
+
+    for hostname in ldap_servers - _valid_ec2_instances():
+        admin_srv.delete(hostname)
+        _LOGGER.info('%s removed from LDAP', hostname)
+    _LOGGER.info('LDAP cleanup complete')
+
+
+def ipa():
+    """Remove terminated ec2 instances from IPA.
+    There shouldn't be any race condition since new instances enter pending
+    status and are therefore part of _VALID_INSTANCES.
+    """
+    _LOGGER.info('IPA cleanup started')
+    ipa_client = _ipa_client()
+    if not ipa_client:
+        return
+    try:
+        ipa_servers = set(ipa_client.get_hosts())
+    except SSLError:
```

Comment:
no need to catch ssl error. this is part of long running task, so in case of error i rather have it crash.

Reply:
removed